client/files.go

```diff
@@ -0,0 +1,188 @@
+	// be included in the call results.
+	Pattern string
+
+	// Itself, when set, will force directory entries not to be listed, but
+	// instead have their information returned as if they were regular files.
+	Itself bool
+}
+
+type FileInfo struct {
+	name    string
+	size    int64
+	mode    fs.FileMode
+	modTime time.Time
+
+	// Path is the full absolute path of the file
+	Path string
+	// UserID is the ID of the owner user
+	UserID int
+	// GroupID is the ID of the owner group
+	GroupID int
+	// User is the string representing the owner user name
+	User string
+	// Group is the string representing the owner user group
+	Group string
+}
+
+// Name returns the base name of the file
+func (fi FileInfo) Name() string {
+	return fi.name
+}
+
+// Size returns the length in bytes for regular files. For others, its
+// behavior is system-dependent
+func (fi FileInfo) Size() int64 {
+	return fi.size
+}
+
+// Mode returns the file mode and permission bits
+func (fi FileInfo) Mode() fs.FileMode {
+	return fi.mode
+}
+
+// ModTime returns the file modification time
+func (fi FileInfo) ModTime() time.Time {
+	return fi.modTime
+}
+
+// IsDir is an abbreviation for Mode().IsDir()
+func (fi FileInfo) IsDir() bool {
+	return fi.mode.IsDir()
+}
+
+// Sys returns the underlying data source (can return nil)
+func (fi FileInfo) Sys() interface{} {
+	return nil
+}
+
+// ListFiles obtains the contents of a directory or glob, or information about a file.
+func (client *Client) ListFiles(opts *ListFilesOptions) ([]*FileInfo, error) {
+	q := make(url.Values)
+	q.Set("action", "list")
+	q.Set("path", opts.Path)
+	if opts.Pattern != "" {
+		q.Set("pattern", opts.Pattern)
+	}
+	if opts.Itself {
+		q.Set("itself", "true")
+	}
+
+	var results []fileInfoResult
+	_, err := client.doSync("GET", "/v1/files", q, nil, nil, &results)
+	if err != nil {
+		return nil, err
+	}
+
+	infos := make([]*FileInfo, len(results))
+	for i, result := range results {
+		infos[i], err = resultToFileInfo(result)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return infos, nil
+}
+
+type fileInfoResult struct {
+	Path         string `json:"path"`
+	Name         string `json:"name"`
+	Type         string `json:"type"`
+	Size         *int64 `json:"size,omitempty"`
+	Permissions  string `json:"permissions"`
+	LastModified string `json:"last-modified"`
+	UserID       *int   `json:"user-id"`
+	User         string `json:"user"`
+	GroupID      *int   `json:"group-id"`
+	Group        string `json:"group"`
+}
+
+func calculateFileMode(fileType string, permissions string) (mode os.FileMode, err error) {
+	p, err := strconv.ParseUint(permissions, 8, 32)
+	if err != nil {
+		return 0, err
+	}
+
+	mode = os.FileMode(p)
+	switch fileType {
+	case "directory":
+		mode |= os.ModeDir
+	case "symlink":
+		mode |= os.ModeSymlink
+	case "socket":
+		mode |= os.ModeSocket
+	case "named-pipe":
+		mode |= os.ModeNamedPipe
+	case "device":
+		mode |= os.ModeDevice
+	default:
+		mode |= os.ModeIrregular
```

Comment:
This may have partly been me being unclear, but I think we still need a `case "file"` that sets no bits, otherwise `"file"` will become `os.ModeIrregular`. I'd do like so:

```go
switch fileType {
case "file":
    // no mode bits set for regular file
case "directory":
    mode |= os.ModeDir
...
default:
    mode |= os.ModeIrregular
}
```

Reply:
Further testing revealed that issue rather quickly, thank you!